<center>
<img src="../../img/ods_stickers.jpg">
    
## [mlcourse.ai](https://mlcourse.ai) - Open Machine Learning Course

Author: [Yury Kashnitsky](https://www.linkedin.com/in/festline/). Translated and edited by [Sergey Isaev](https://www.linkedin.com/in/isvforall/), [Artem Trunov](https://www.linkedin.com/in/datamove/), [Anastasia Manokhina](https://www.linkedin.com/in/anastasiamanokhina/), and [Yuanyuan Pao](https://www.linkedin.com/in/yuanyuanpao/). All content is distributed under the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

**In this task you should use Pandas to answer a few questions about the [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) dataset. (You don't have to download the data – it's already  in the repository)**

Unique values of all features (for more information, please see the links above):
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max.columns', 100)
# to draw pictures in jupyter notebook
%matplotlib inline 
import matplotlib.pyplot as plt
import seaborn as sns
# we don't like warnings
# you can comment the following 2 lines if you'd like to
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('../../data/adult.data.csv')
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. How many men and women (*sex* feature) are represented in this dataset?** 

In [3]:
# Basic profiling of dataset 
data.sex.value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

# Answer 1: 
#### Count of Men in the dataset   = 21790
#### Count of Women in the dataset   = 10771

**2. What is the average age (*age* feature) of women?**

In [4]:
data[data.sex=='Female'].age.mean()

36.85823043357163

# Answer 2: 
#### Average age of women : 36.85823043357163

**3. What is the percentage of German citizens (*native-country* feature)?**

In [5]:
# Finding Count of German Citizens
gc=data[data['native-country']=='Germany']['native-country'].count()

In [6]:
# Finding Total count of Dataset
tc=data.age.count()

In [7]:
# Percentage of German Citizens = (Finding Count of German Citizens)/(Finding Total count of Dataset)*100
PGC=gc/tc*100

In [8]:
print(PGC)

0.42074874850281013


# Answer 3: 
####  Percentage of German Citizens = _0.42074874850281013_



**4-5. What are the mean and standard deviation of age for those who earn more than 50K per year (*salary* feature) and those who earn less than 50K per year? **

In [9]:
# Dataframe of people with salary greater than 50K

gt5k=data[data.salary=='>50K']
print( 'Mean Age of people with salary greater than 50K ', gt5k.age.mean())
print( 'Standard Deviation of Age of people with salary greater than 50K ', gt5k.age.std())

Mean Age of people with salary greater than 50K  44.24984058155847
Standard Deviation of Age of people with salary greater than 50K  10.51902771985177


In [10]:
# Dataframe of people with salary less than 50K
lt5k=data[data.salary=='<=50K']

print( 'Mean Age of people with salary less than 50K ', lt5k.age.mean())
print( 'Standard Deviation of Age of people with salary less than 50K ', lt5k.age.std())


Mean Age of people with salary less than 50K  36.78373786407767
Standard Deviation of Age of people with salary less than 50K  14.020088490824813


# Answer 4-5: 
#### Mean Age of people with salary greater than 50K                              : 44.24984058155847
#### Standard Deviation of Age of people with salary greater than 50K  : 10.51902771985177

#### Mean Age of people with salary less than 50K                                   : 36.78373786407767
#### Standard Deviation of Age of people with salary less than 50K       : 14.020088490824813

_seems like older folks make more money, they might have wisdom to impart :)_


**6. Is it true that people who earn more than 50K have at least high school education? (*education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate* feature)**

In [11]:
# Profiling the education series for people with more than 50K Salary
gt5k.education.value_counts()

Bachelors       2221
HS-grad         1675
Some-college    1387
Masters          959
Prof-school      423
Assoc-voc        361
Doctorate        306
Assoc-acdm       265
10th              62
11th              60
7th-8th           40
12th              33
9th               27
5th-6th           16
1st-4th            6
Name: education, dtype: int64

# Answer 6: 
#### Not true. 
#### Not all people with salaries over 50K have completed high school education. 
#### We have people starting from only completing 1st-4th up earning over 50K. The validity of the dataset needs to be ascertained though.

**7. Display age statistics for each race (*race* feature) and each gender (*sex* feature). Use *groupby()* and *describe()*. Find the maximum age of men of *Amer-Indian-Eskimo* race.**

# Answer 7 a
####  Display age statistics for each race using groupby() and describe()

In [12]:
# Display age stats for each race ( )
data.groupby('race')['age'].describe()

count       mean        std   min   25%   50%   75%  \
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                     max  
race                      
Amer-Indian-Eskimo  82.0  
Asian-Pac-Islander  90.0  
Black               90.0  
Other               77.0  
White               90.0

In [13]:
# Display age stats for each race ( ) -- Alternative method to compute (ignored as per question)
# data.groupby('race').agg({'age':['count', 'min','max', 'mean', 'median',  'std' ]})

# Answer 7 b
####  Display age statistics for each gender  using groupby() and describe()

In [14]:
# Display age stats for each Gender ( )
data.groupby('sex')['age'].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                                
Female  10771.0  36.858230  14.013697  17.0  25.0  35.0  46.0  90.0
Male    21790.0  39.433547  13.370630  17.0  29.0  38.0  48.0  90.0

# Answer 7 c
####  Find maximum age of men of *Amer-Indian-Eskimo* race

In [15]:
# Maximum Age of men of Amer-Indian-Eskimo race
Max_age_of_AIE_Men=data[(data.race=='Amer-Indian-Eskimo') & (data.sex=='Male')].age.max()
print('The maximum age of Amer-Indian-Eskimo men is : ',  Max_age_of_AIE_Men)

The maximum age of Amer-Indian-Eskimo men is :  82


**8. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (*marital-status* feature)? Consider as married those who have a *marital-status* starting with *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.**

In [16]:
# Total Count of Married men with Salary over 50K
men=gt5k[(gt5k['sex']=='Male')]
total_men_count=men.shape[0]
print(total_men_count)

6662


In [17]:
mmc=men[(men['marital-status']=='Married-civ-spouse') | 
        (men['marital-status']=='Married-spouse-absent') | 
        (men['marital-status']=='Married-AF-spouse')]['marital-status']
total_married_men_count=len(mmc)
print(total_married_men_count)

5965


In [18]:
pc=total_married_men_count/total_men_count*100
print('Percentage of married men earning more than 50K is ', pc)

Percentage of married men earning more than 50K is  89.53767637346142


# Answer 8
####  An overwhelming 89.53% of the total men earning over 50K salary is married


**9. What is the maximum number of hours a person works per week (*hours-per-week* feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?**

# Answer 9 a
####  Maximum number of hours a person works per week : 99

In [19]:
# max no of hours a person works per week
max_hour_hour = data['hours-per-week'].max()
print(max_hour_hour)


99


# Answer 9 b
####  Number of people that work maximum number of hours a person works per week in the dataset: 85

In [20]:
# How many people work max no of hours per week
ppl_working_max_hours=data[(data['hours-per-week']== max_hour_hour)]
count_ppl_working_max_hours= ppl_working_max_hours.shape[0]
print(count_ppl_working_max_hours)

85


# Answer 9 c
####  Percentage of high earners who work maximum number of hours per week in the dataset : 29.4

In [21]:
# What % of those earn > 50K
high_earners_working_max_hours=ppl_working_max_hours[ppl_working_max_hours.salary=='>50K'].shape[0]
pc_of_high_earners_working_max=high_earners_working_max_hours/count_ppl_working_max_hours*100
print(pc_of_high_earners_working_max) 

29.411764705882355


**10. Count the average time of work (*hours-per-week*) for those who earn a little and a lot (*salary*) for each country (*native-country*). What will these be for Japan?**

# Answer 10b
#### Average time of work for those who earn a little and a lot for Japan

In [23]:
#  Average time of work for those who earn a little and a lot for Japan
data[data['native-country']=='Japan'].groupby(['native-country','salary']).agg({'hours-per-week':['min','max']}).unstack()

hours-per-week                
                          min        max     
salary                  <=50K >50K <=50K >50K
native-country                               
Japan                      10   21    65   99

# Answer 10a
#### Average time of work for those who earn a little and a lot for each country

In [26]:
# Average time of work for those who earn a little and a lot for each country
data.groupby(['native-country','salary']).agg({'hours-per-week':['mean']}).unstack()

hours-per-week           
                                     mean           
salary                              <=50K       >50K
native-country                                      
?                               40.164760  45.547945
Cambodia                        41.416667  40.000000
Canada                          37.914634  45.641026
China                           37.381818  38.900000
Columbia                        38.684211  50.000000
Cuba                            37.985714  42.440000
Dominican-Republic              42.338235  47.000000
Ecuador                         38.041667  48.750000
El-Salvador                     36.030928  45.000000
England                         40.483333  44.533333
France                          41.058824  50.750000
Germany                         39.139785  44.977273
Greece                          41.809524  50.625000
Guatemala                       39.360656  36.666667
Haiti                           36.325000  42.750000
Holand-Netherlands              40.000000        NaN
Honduras                        34.333333  60.000000
Hong                            39.142857  45.000000
Hungary                         31.300000  50.000000
India                           38.233333  46.475000
Iran                            41.440000  47.500000
Ireland                         40.947368  48.000000
Italy                           39.625000  45.400000
Jamaica                         38.239437  41.100000
Japan                           41.000000  47.958333
Laos                            40.375000  40.000000
Mexico                          40.003279  46.575758
Nicaragua                       36.093750  37.500000
Outlying-US(Guam-USVI-etc)      41.857143        NaN
Peru                            35.068966  40.000000
Philippines                     38.065693  43.032787
Poland                          38.166667  39.000000
Portugal                        41.939394  41.500000
Puerto-Rico                     38.470588  39.416667
Scotland                        39.444444  46.666667
South                           40.156250  51.437500
Taiwan                          33.774194  46.800000
Thailand                        42.866667  58.333333
Trinadad&Tobago                 37.058824  40.000000
United-States                   38.799127  45.505369
Vietnam                         37.193548  39.200000
Yugoslavia                      41.600000  49.500000